### PR TITLE
drivers, server: support webhook-defined OS, OS URLs, S3-compatible stores

### LINKED ISSUE #1263: Spec: webhook specified object stores
## Abstract
We should allow to provide object store credentials in web hook response.

## Motivation
Right now object store credentials is specified per broadcaster, we want to specify it per stream.

## Proposed Solution
Web hook response should send web hook credentials to the broadcaster.

## Implementation Tasks and Considerations
Eli Edit: This will follow the fields and syntax defined here https://github.com/livepeer/go-livepeer/issues/1572

## Additional Context
We should also add authentication for web hook, but this is out of scope for this spec.

### LINKED ISSUE #1572: proposal: generic URL-based OS syntax
## Abstract
Let's define OS like `s3://whatever`

## Motivation

Currently, OS have to be defined the `-s3bucket`, `-s3creds`, and other such CLI options. If we continue to add more CLI parameters for each OS, we'll have thousands. Additionally, we're going to want to refer to OS for failed segments (#1398), recording (#1565) and webhooks at some point, so it's good to have a generic way to refer to OSs.

## Proposed Solution

TLDR: Change OS to have the form s3://ACCESS_KEY_ID:SECRET_ACCESS_KEY@eu-central-1/testbucket. Those get called by `-objectStore`, and maybe eventually `-failObjectStore`, and `-recordObjectStore`.

This will allow us to refer to object stores generically in multiple contexts using the same syntax. We should probably deprecate and support the previous options for backwards-compatibilty. This proposed syntax borrows heavily from ffmpeg's syntax for customizing various protocols.

### S3 stores

Before:

```shell
livepeer -broadcaster -s3bucket eu-central-1/testbucket -s3creds ACCESS_KEY_ID/SECRET_ACCESS_KEY
```

After:

```
livepeer -broadcaster -objectStore s3://ACCESS_KEY_ID:SECRET_ACCESS_KEY@eu-central-1/testbucket
```

### GCP Stores

This one allows a keyfile, so I'm not really sure how to express that. This is what you'd technically do for URL compatibility with `./path/to/keyfile` but it sure isn't very fun to write on the command line:

```
livepeer -broadcaster -objectStore "gs://ACCESS_KEY_ID:.%2Fpath%2Fto%2Fkeyfile@example-gcp-bucket/path"
```

We should also support providing the key as an inline JSON blob, but that's even harder to type.

### Custom S3 Stores (https://github.com/livepeer/go-livepeer/pull/1206)

Before:

```
livepeer -broadcaster -s3bucket http://localhost:9000/eli-bucket -s3creds minioadmin/minioadmin
```

After:

```shell
livepeer -broadcaster -objectStore s3://minioadmin:minioadmin@localhost:9000/eli-bucket
```

### Saving errored segments (https://github.com/livepeer/go-livepeer/pull/1398)

```
livepeer -orchestrator -transcoder -failObjectStore s3://ACCESS_KEY_ID:SECRET_ACCESS_KEY@eu-central-1/testbucket
```

### Filesystem storage (https://github.com/livepeer/go-livepeer/commit/5b776565debc43256a1ad235896f89faebd79a86)

```
livepeer -broadcaster -objectStore file://./path/to/whatever
```

(Or file could be the default to make it a little cleaner.)

## Implementation Tasks and Considerations

We should probably do the primary OS refactor first, ship that, and then incorporate long-running PRs like https://github.com/livepeer/go-livepeer/pull/1398 and https://github.com/livepeer/go-livepeer/pull/1206.

## Known Unknowns

How do we handle GCP? Their access keys are massive JSON blobs containing a private TLS key. Separate keyfile param as a query string?

## Alternatives

Perhaps instead of having three things called `objectStore`, `failObjectStore`, and `recordObjectStore`, we could add support for multiple object stores per-stream, with each one configurable to handle different cases...

```
livepeer -broadcaster -objectStore s3://ACCESS_KEY_ID:SECRET_ACCESS_KEY@eu-central-1/testbucket -objectStore file://./my-recordings?record=true -objectStore gs://user:password/bucket/failed-segments?failed=true
```

The only use case I can think of for that would be to have multiple records... you can't have more than one primary OS and I'm not sure why you'd want to send failed segments several places. Multiple records, though, is potentially interesting — perhaps you want to record to an S3 bucket but also save a copy on the local filesystem as a backup.

### PR DESCRIPTION
**Pull request by iameli**
**What does this pull request do? Explain your changes. (required)**

* URL-based OS refactor: https://github.com/livepeer/go-livepeer/issues/1572
* Webhook-defined OS https://github.com/livepeer/go-livepeer/issues/1263
* S3-compatible OS support https://github.com/livepeer/go-livepeer/pull/1206

**Specific updates (required)**

- Added `-objectStore` parameter per #1572 
- `-s3creds`, `-s3bucket`, `-gskey`, and `-gsbucket` now display a deprecated message and prompt with the correct OS URL
- Added `s3+http://` and `s3+https://` syntax for supporting OS outside of the primary S3 store
- Added an `objectStorage` 

Still to do:
- Figure out how to handle the IsOwnExternal workflow, that'll break with hook-defined OS at the moment
- Update webhook docs
- Integrate with record mode after that lands (or merge these PRs together)
- GS support is a mess
  - `-objectStore gs://%2Fpath%2Fto%2Fkey@bucket-name` is ugly and hard to type
  - There's no actual way to define GS credentials through the webhook, you can only provide them through the command line. The good news is I did test embedding an entire JSON blob in the username field of the URL and it parsed totally fine, so we can just add that for webhook support.
  - On the other hand, we should probably _dis_allow the path option from the webhook. I mean, generally we trust whoever is on the opposite end of `-authWebhookUrl`, but that's no reason to give them the ability to probe for the existence of arbitrary files on the local filesystem!

**How did you test each of these updates (required)**
So far I've just tested with Minio locally. (I'm on the streaming laptop, so no actual cloud credentials allowed here.) I'll make sure to test with S3 & Minio & GCP on both the B and the O before this ships

**Does this pull request close any open issues?**

Fixes #1263
Fixes #1206 
Fixes #1572 

**Checklist:**

- [ ] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

### CODE REVIEW

**Comment by darkdarkdragon** [comment]
Looks good

> On the other hand, we should probably _dis_allow the path option from the webhook. 

Sounds reasonable

> Figure out how to handle the IsOwnExternal workflow, that'll break with hook-defined OS at the moment

Best option I think would be:
to `OSSession` interface add method `IsOwn(uri string) bool`
In `broadcast.go` change
`if verifier != nil || (bos != nil && !drivers.IsOwnExternal(url)) {`
to
`if verifier != nil || (bos != nil && !bos.IsOwn(url)) {`

In `orchestrator.go` `if drivers.IsOwnExternal(seg.Name)` - left as is

In `verify.go` to `type Params struct` add `BroadcasterOS    drivers.OSSession` field and in `epic.go` change
`if drivers.IsOwnExternal(params.Source.Name) {` to
`if  params.BroadcasterOS != nil && params.BroadcasterOS.IsOwn(params.Source.Name) {`

Other option would be to keep global map of all OS drivers created through web hooks and have `IsOwnExternal` to check against that map - but that will require some cleanup logic.

**Comment by iameli** [comment]
@darkdarkdragon Thanks, on this now.

I'm thinking of the GS file syntax like this: `-objectStore gs://bucket-name?keyfile=/path/to/key`. This pattern would also let us eventually do things like `-objectStore s3://bucket-name?keyfile=/home/iameli/.aws/credentials`, which is kinda nice.

Then, in the CLI, we run a "PrepareOSURL" step that transforms `gs://bucket-name?keyfile=/path/to/key` into `gs://< insert ugly URL-encoded JSON blob here>@bucket-name`.

In the webhook, we won't run that function, so the file options will be a no-op.

**Comment by iameli** [comment]
How do I know whether one of these strings IsOwn? Is it just that the server hostnames and paths match, implying they're the same bucket?

**Comment by iameli** [comment]
Ok, I think I have most of the IsOwn stuff in place. I'm not sure how to get `IsOwn` in epic.go, maybe I'm missing something — looks like the local struct doesn't have an OS.

**Comment by iameli** [comment]
Closing in favor of https://github.com/livepeer/go-livepeer/pull/1582
